### PR TITLE
feat: sgx backend sends CRL for attestation

### PR DESCRIPTION
**Pull request by rjzak**
Refs:
* https://github.com/enarx/enarx/issues/2389
* https://github.com/profianinc/steward/issues/112

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

**Comment by rjzak** [comment]
Some tests were failing since compiling for `x86_64-unknown-linux-musl` didn't like `der::Decode`, and wanted `x509_cert::der::Decode`, despite 1) it working for GNU, and 2) it's still the same version of he `der` crate used in both instances.

Another issue is Musl not liking `anyhow::Context`, which seems to be Musl not picking up on the `der::Encode` trait returning a `Result<Vec<u8>>`, not just a `Vec<u8>`.

File: src/backend/sgx/attestation.rs [hunk 1]
```diff
@@ -434,6 +442,62 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+    let q_size = get_quote_size(akid.clone())?;
+    let mut out_buf_temp = Vec::with_capacity(q_size);
+    let report = [0u8; SGX_REPORT_SIZE];
+    get_quote(&report, akid, &mut out_buf_temp)?;
+
+    let mut crl_file = sgx_cache_dir()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX CRL read error {}", e)))?;
+    crl_file.push("crls.der");
+
+    let evidence = SgxEvidence {
```
Comment (rjzak) [inline src/backend/sgx/attestation.rs:456]: @haraldh @dpal what's better, a `struct` that we convert to DER, or just add the CRL bytes to the end of the report bytes? Just adding the bytes might be easier, since the CRL bytes are already DER-encoded, and saves the issue ot figuring how how big the report is once made into DER.

Secondly, the `exec-wasmtime` already converts the report to ASN1 when making the extension, so we probably don't want two levels of ASN1 encoding.

The only issue is making sure we get the remaining (CRL) bytes after Rust converts the bytes to the C-style struct.

File: tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs [hunk 2]
```diff
@@ -111,6 +122,10 @@ fn main() -> std::io::Result<()> {
 
     assert!(matches!(tech, TeeTech::Sgx));
 
+    let evidence = SgxEvidence::from_der(&buffer)
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX evidence to DER error {}", e)))?;
+    let buffer = evidence.quote;
+
     let bytes = &buffer[MRSIGNER_START..][..32];
     let out = hex::encode(bytes);
     println!("MRSIGNER = {}", out);
```
Comment (rjzak) [inline tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs:131]: @haraldh This works:
* Decoding from ASN1
* Getting `MRSIGNER`

This doesn't work:
* Getting the quote sig start: `let bytes = &buffer[QUOTE_SIG_START..len];`. `QUOTE_SIG_START` = 436, but this seems to be out of range.
```
MRSIGNER = 0b33f6e61130793039bcda31ca4b316c744f983ff0a72a2035f496b8eef29106
thread 'main' panicked at 'range end index 7869 out of range for slice of length 4600', tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs:133:18
```

Thoughts?

File: tests/crates/enarx_syscall_tests/src/bin/sgx_get_att_quote.rs [hunk 3]
```diff
@@ -45,13 +45,14 @@ fn main() -> Result<()> {
...
+        return Err(2);
     }
 
-    /* check beginning of quote matches expected value */
-    if !buf.starts_with(&expected) {
-        return Err(1);
-    }
+    /* check beginning of quote matches expected value, after the ASN.1 header */
+    //if !buf[0..8].starts_with(&expected) {
+    //    // New expected value seems to be 030002000000000009000d00939a7233f79c4ca9940a0db3957f0607
```
Comment (rjzak) [inline tests/crates/enarx_syscall_tests/src/bin/sgx_get_att_quote.rs:53]: @haraldh Why the change in expected value? Is this a hardware thing? It's only slightly off.

Comment (dpal) [reply tests/crates/enarx_syscall_tests/src/bin/sgx_get_att_quote.rs:53]: How off is it? Can it be that a part of the encoding blends into the beginning of the report? Which bytes are different?

File: tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs (hunk 2 above)
Comment (haraldh) [reply tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs:131]: please! simple fix... as the error says `line 133`...

```rust
    let bytes = &buffer[QUOTE_SIG_START..];
```

`len` is not the length of the quote anymore.. it's the overall length.

**Review by haraldh — CHANGES_REQUESTED** [review]

File: tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs (hunk 2 above)
Comment (rjzak) [reply tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs:138]: Super!

File: src/backend/sgx/attestation.rs [hunk 4]
```diff
@@ -434,6 +438,36 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+/// DER-encoded list of CRLs
+pub fn retrieve_crls() -> anyhow::Result<Vec<u8>> {
+    #[derive(Sequence)]
+    struct CrlList<'a> {
+        crls: Vec<CertificateList<'a>>,
+    }
+
+    let sgx_cache = sgx_cache_dir().context("unable to get SGX cache directory")?;
+    let paths =
+        std::fs::read_dir(sgx_cache).context("getting SGX CRLs from /var/cache/intel-sgx")?;
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: This code was outdated, the CRLs are now in one file, `/var/cache/intel-sgx/crls.der`. No old files to clean up.

File: src/backend/sgx/attestation.rs [hunk 5]
```diff
@@ -11,20 +11,30 @@ use crate::protobuf::aesm_proto::{
...
 
 const SGX_TI_SIZE: usize = size_of::<TargetInfo>();
 
 const AESM_REQUEST_TIMEOUT: u32 = 1_000_000;
 const SGX_KEY_ID_SIZE: u32 = 256;
 const SGX_REPORT_SIZE: usize = 432;
 
+#[derive(Sequence)]
+pub struct SgxEvidence {
+    pub aesm: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Outdated, as it was already named `quote`.

File: src/backend/sgx/attestation.rs [hunk 6]
```diff
@@ -11,20 +11,30 @@ use crate::protobuf::aesm_proto::{
...
+    pub crl: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: It's not just the CRL, it's also the report. The SNP case calls the SNP and vcek "Evidence". I was trying to be consistent.

File: src/backend/sgx/enarxcall.rs [hunk 7]
```diff
@@ -216,9 +218,11 @@ pub(crate) fn sgx_enarxcall<'a>(
             let akid = get_attestation_key_id().context(
                 "Error obtaining attestation key id. Check your aesmd / pccs service installation.",
             )?;
-            *ret = get_quote_size(akid).context(
-                "Error getting quote size. Check your aesmd / pccs service installation.",
-            )?;
+            *ret = get_quote_size_with_collateral(akid)
+                //*ret = get_quote_size(akid)
```
Comment (rvolosatovs) [inline src/backend/sgx/enarxcall.rs:222]: Why is this here?

File: src/backend/sgx/enarxcall.rs [hunk 8]
```diff
@@ -202,7 +203,8 @@ pub(crate) fn sgx_enarxcall<'a>(
             let akid = get_attestation_key_id().context(
                 "Error obtaining attestation key id. Check your aesmd / pccs service installation.",
             )?;
-            *ret = get_quote(report_buf, akid, quote_buf)
+            *ret = get_quote_and_collateral(report_buf, akid, quote_buf)
+                //*ret = get_quote(report_buf, akid, quote_buf)
```
Comment (rvolosatovs) [inline src/backend/sgx/enarxcall.rs:207]: why commented-out code?

File: src/backend/sgx/attestation.rs [hunk 9]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+
+    let mut crl_file = sgx_cache_dir()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX CRL read error {e}")))?;
+    crl_file.push("crls.der");
+
+    let evidence = SgxEvidence {
+        quote: &out_buf_temp.clone(),
+        crl: Document::read_der_file(&crl_file)
+            .context(format!("error reading Intel CRL file {crl_file:?}"))
+            .unwrap(),
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:488]: Don't panic in production code

File: src/backend/sgx/attestation.rs [hunk 10]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+    get_quote(report, akid, &mut out_buf_temp)?;
+
+    let mut crl_file = sgx_cache_dir()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX CRL read error {e}")))?;
+    crl_file.push("crls.der");
+
+    let evidence = SgxEvidence {
+        quote: &out_buf_temp.clone(),
+        crl: Document::read_der_file(&crl_file)
+            .context(format!("error reading Intel CRL file {crl_file:?}"))
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:487]:
```suggestion
            .context(format!("error reading Intel CRL file `{crl_file:?}`"))
```

File: src/backend/sgx/attestation.rs [hunk 11]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+    akid: Vec<u8>,
+    out_buf: &mut [u8],
+) -> Result<usize, Error> {
+    // Get the report in a separate buffer
+    let q_size = get_quote_size(akid.clone())?;
+    let mut out_buf_temp = vec![0; q_size];
+    get_quote(report, akid, &mut out_buf_temp)?;
+
+    let mut crl_file = sgx_cache_dir()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX CRL read error {e}")))?;
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:481]:
```suggestion
        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX CRL read error: {e}")))?;
```

File: src/backend/sgx/attestation.rs [hunk 12]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+    let evidence = SgxEvidence {
+        quote: &out_buf_temp,
+        crl: Document::read_der_file(&crl_file)
+            .context(format!("error reading Intel CRL file {crl_file:?}"))
+            .unwrap(),
+    };
+
+    let evidence_vec = evidence
+        .to_vec()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX evidence to DER error {e}")))?;
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:465]:
```suggestion
        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX evidence to DER error: {e}")))?;
```

File: src/backend/sgx/attestation.rs [hunk 13]
```diff
@@ -4,27 +4,37 @@
 // * https://github.com/fortanix/rust-sgx for examples of AESM requests.
 
 use super::AESM_SOCKET;
+use crate::backend::sgx::sgx_cache_dir;
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:8]:
```suggestion
```

File: src/backend/sgx/attestation.rs [hunk 14]
```diff
@@ -4,27 +4,37 @@
...
 use crate::protobuf::aesm_proto::{
     Request, Request_GetQuoteExRequest, Request_GetQuoteSizeExRequest,
     Request_GetSupportedAttKeyIDNumRequest, Request_GetSupportedAttKeyIDsRequest,
     Request_InitQuoteExRequest, Response,
 };
 
+use anyhow::Context;
+use der::{Document, Encode, Sequence};
+use protobuf::Message;
+use sallyport::item::enarxcall::sgx::TargetInfo;
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:18]: This is not correctly placed

File: src/backend/sgx/attestation.rs [hunk 15]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
     Ok(quote.len())
 }
 
+/// Gets quote size with CRL added
+pub fn get_quote_size_with_collateral(akid: Vec<u8>) -> Result<usize, Error> {
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:448]:
```suggestion
pub fn get_quote_size_with_collateral(akid: Vec<u8>) -> io::Result<usize> {
```

File: src/backend/sgx/attestation.rs [hunk 16]
```diff
@@ -434,6 +444,73 @@ pub fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usi
...
+        .map_err(|e| Error::new(ErrorKind::Other, format!("SGX evidence to DER error {e}")))?;
+
+    Ok(evidence_vec.len())
+}
+
+pub fn get_quote_and_collateral(
+    report: &[u8],
+    akid: Vec<u8>,
+    out_buf: &mut [u8],
+) -> Result<usize, Error> {
```
Comment (rvolosatovs) [inline src/backend/sgx/attestation.rs:474]:
```suggestion
) -> io::Result<usize> {
```

File: src/backend/sgx/data.rs [hunk 17]
```diff
@@ -172,3 +176,83 @@ pub fn aesm_socket() -> Datum {
...
+
+    let crls =
+        match std::fs::read(crl_file) {
+            Ok(c) => c,
+            Err(e) => return Datum {
+                name,
+                pass: false,
+                info: Some(e.to_string()),
+                mesg: Some(
+                    "Re-run `enarx platform sgx cache-crl` to generate the Intel CRL cache file"
```
Comment (rvolosatovs) [inline src/backend/sgx/data.rs:216]: This should specify that reading the file failed and why

File: src/backend/sgx/data.rs [hunk 18]
```diff
@@ -172,3 +176,83 @@ pub fn aesm_socket() -> Datum {
...
+
+    let crls =
+        match Vec::<CertificateList>::from_der(&crls) {
+            Ok(c) => c,
+            Err(e) => return Datum {
+                name,
+                pass: false,
+                info: Some(e.to_string()),
+                mesg: Some(
+                    "Re-run `enarx platform sgx cache-crl` to generate the Intel CRL cache file"
```
Comment (rvolosatovs) [inline src/backend/sgx/data.rs:230]: This should specify that parsing the file failed and why

File: src/backend/sgx/attestation.rs (hunk 16 above)
Comment (rjzak) [reply src/backend/sgx/attestation.rs:474]: This would make the return type inconsistent with the return type in the other functions in `attestation.rs`.

File: src/backend/sgx/data.rs (hunk 17 above)
Comment (rjzak) [reply src/backend/sgx/data.rs:216]: The `info` field gets the error message as on other examples with `Datum`, isn't that the correct place? I did add the path for the file in question.

File: src/backend/sgx/data.rs (hunk 18 above)
Comment (rjzak) [reply src/backend/sgx/data.rs]: The `info` field gets the error message as on other examples with Datum, isn't that the correct place? I did add the path for the file in question.

File: src/backend/sgx/attestation.rs (hunk 14 above)
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Resolved according to how the imports were sorted in `enarxcall.rs`

File: src/backend/sgx/attestation.rs (hunk 12 above)
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Fixed

File: src/backend/sgx/attestation.rs (hunk 11 above)
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Fixed

File: src/backend/sgx/attestation.rs (hunk 10 above)
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Fixed

File: src/backend/sgx/attestation.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Done

Comment (rjzak) [reply src/backend/sgx/attestation.rs]: Good catch, fixed!

File: src/backend/sgx/enarxcall.rs (hunk 8 above)
Comment (rjzak) [reply src/backend/sgx/enarxcall.rs]: Forgot that. Removed.

File: src/backend/sgx/enarxcall.rs (hunk 7 above)
Comment (rjzak) [reply src/backend/sgx/enarxcall.rs]: Forgot that, removed.

File: crates/sallyport/src/item/enarxcall/mod.rs [hunk 19]
```diff
@@ -60,29 +63,35 @@ pub enum Number {
     /// SGX quote size request call number.
     GetSgxQuoteSize = 0x05,
 
+    /// SGX CRLs request
+    GetSGXCrls = 0x06,
```
Comment (rvolosatovs) [inline crates/sallyport/src/item/enarxcall/mod.rs:67]:
```suggestion
    GetSgxCrls = 0x06,
```

File: crates/sallyport/src/item/enarxcall/mod.rs [hunk 20]
```diff
@@ -60,29 +63,35 @@ pub enum Number {
     /// SGX quote size request call number.
     GetSgxQuoteSize = 0x05,
 
+    /// SGX CRLs request
```
Comment (rvolosatovs) [inline crates/sallyport/src/item/enarxcall/mod.rs:66]: If we are reordering these, let's come up with a convention and group things close?

For example, we could use:
- 0x00 - 0x0F for all SGX calls
- 0x10 - 0x1F for all SNP calls
- 0x20 - 0x9F reserved for future technologies
- 0xA0 - 0xFF for all other calls

**Review by rvolosatovs — COMMENTED** [review]
Did not do a thorough review, just a quick look on my phone

Good job on the Sallyport calls!

File: crates/sallyport/src/item/enarxcall/mod.rs (hunk 20 above)
Comment (rjzak) [reply crates/sallyport/src/item/enarxcall/mod.rs:66]: That would be good. I also noticed that the numbering went from `0x08`, `0x09`, ... `0x10`, `0x11`, so I fixed that.

File: crates/exec-wasmtime/src/runtime/identity/mod.rs [hunk 21]
```diff
@@ -91,12 +92,33 @@ pub fn generate() -> anyhow::Result<(Zeroizing<Vec<u8>>, Vec<u8>)> {
...
 
     let attestation_report = platform.attest(&key_hash)?;
+    let crls = platform.crls()?;
+
+    //eprintln!("Exec: CLRs: {:?}", crls);
+
+    //let mut evidence = vec![0x30u8, 0x02u8];
+    //evidence.append(&mut attestation_report);
+    //evidence.append(&mut crls);
+
```
Comment (haraldh) [inline crates/exec-wasmtime/src/runtime/identity/mod.rs]:
```suggestion
```

File: tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs [hunk 22]
```diff
@@ -112,15 +122,30 @@ fn main() -> io::Result<()> {
 
     let mut buffer = vec![0u8; len];
 
-    let (len, tech) = get_att_syscall(Some(&mut nonce[..]), Some(&mut buffer))?;
+    let (_len, tech) = get_att_syscall(Some(&mut nonce[..]), Some(&mut buffer))?;
 
     assert!(matches!(tech, TeeTech::Sgx));
 
+    let evidence = SgxEvidence::from_der(&buffer).map_err(|e| {
```
Comment (haraldh) [inline tests/crates/enarx_exec_tests/src/bin/sgx_attestation.rs:129]:
```suggestion
    let evidence = SgxEvidence::from_der(&buffer[..len]).map_err(|e| {
```

**Review by haraldh — APPROVED** [review]
